Source/Particles/ERFPCEvolve.cpp

```diff
@@ -130,10 +129,16 @@ void ERFPC::AdvectWithFlow ( MultiFab*                           a_umac,
                     iv[1] += domain.smallEnd()[1];
                     ParticleReal zlo, zhi;
                     if (use_terrain) {
-                        zlo = Real(0.25) * (zheight(iv[0],iv[1]  ,iv[2]  ) + zheight(iv[0]+1,iv[1]  ,iv[2]  ) +
-                                            zheight(iv[0],iv[1]+1,iv[2]  ) + zheight(iv[0]+1,iv[1]+1,iv[2]  ));
-                        zhi = Real(0.25) * (zheight(iv[0],iv[1]  ,iv[2]+1) + zheight(iv[0]+1,iv[1]  ,iv[2]+1) +
-                                            zheight(iv[0],iv[1]+1,iv[2]+1) + zheight(iv[0]+1,iv[1]+1,iv[2]+1));
+                        Real lx = (p.pos(0)-plo[0])*dxi[0] - static_cast<Real>(iv[0]-domain.smallEnd()[0]);
+                        Real ly = (p.pos(1)-plo[1])*dxi[1] - static_cast<Real>(iv[1]-domain.smallEnd()[1]);
+                        zlo =  zheight(iv[0]  ,iv[1]  ,iv[2]  ) * (1.0-lx) * (1.0-ly) +
+                               zheight(iv[0]+1,iv[1]  ,iv[2]  ) *      lx  * (1.0-ly) +
+                               zheight(iv[0]  ,iv[1]+1,iv[2]  ) * (1.0-lx) * ly +
+                               zheight(iv[0]+1,iv[1]+1,iv[2]  ) *      lx  * ly;
+                        zhi =  zheight(iv[0]  ,iv[1]  ,iv[2]+1) * (1.0-lx) * (1.0-ly) +
+                               zheight(iv[0]+1,iv[1]  ,iv[2]+1) *      lx  * (1.0-ly) +
+                               zheight(iv[0]  ,iv[1]+1,iv[2]+1) * (1.0-lx) * ly +
+                               zheight(iv[0]+1,iv[1]+1,iv[2]+1) *      lx  * ly;
                     } else {
                         zlo =  iv[2]    * dx[2];
                         zhi = (iv[2]+1) * dx[2];
```

Comment:
Can lines 130-145 (or 130-151) be written as a separate function/functor? If functor, it could store `zheight` and `use_terrain` as member data?

Reply:
Actually I want to move this functionality into a routine in amrex -- it is used by multiple codes.   I was just going to make a new function called something like update_particle_idata that takes one particle and the zheight array.   What do you think?

Reply:
That would be nice since this is a particle functionality, not an ERF-specific thing. But terrain and zheight are specific to atmospheric flows. To make it more general, could we have xcoord, ycoord, zcoord (defaulted to nullptr?) instead of just zheight that allows mapping in all dimensions? ERF would then call this function with only zheight.

Reply:
We're working on the generalized functionality ...

Reply:
I think we could parallel what is done in the [BinMapper](https://github.com/AMLattanzi/amrex/blob/2fc8a34ced0c0acae04298c69ec23803ec05e9cd/Src/Particle/AMReX_ParticleUtil.H#L295-L338) functor. This would move the functionality to AMReX, clean up the ERF code since the above snippet now lives in a functor, and would be similar to what we do without curvilinear coordinates.